### ISSUE TITLE
ConcurrentModificationException when destroying Activity

### ISSUE THREAD

**Issue by dmarcato** [bluelinelabs/Conductor#127]
I'm getting randomly this `Exception` when closing an Activity - I think it might happen if I have a long stack of `Controller`s and I quickly press the back button to exit the app.
```
FATAL EXCEPTION: main
Process: com.invoice2go.dev, PID: 26809
java.lang.RuntimeException: Unable to destroy activity {com.invoice2go.dev/com.invoice2go.activity.MainActivity}: java.util.ConcurrentModificationException
    at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3831)
    at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:3849)
    at android.app.ActivityThread.-wrap5(ActivityThread.java)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1398)
    at android.os.Handler.dispatchMessage(Handler.java:102)
    at android.os.Looper.loop(Looper.java:148)
    at android.app.ActivityThread.main(ActivityThread.java:5417)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
Caused by: java.util.ConcurrentModificationException
    at java.util.ArrayList$ArrayListIterator.next(ArrayList.java:573)
    at com.bluelinelabs.conductor.ControllerHostedRouter.removeHost(ControllerHostedRouter.java:50)
    at com.bluelinelabs.conductor.Controller.removeViewReference(Controller.java:790)
    at com.bluelinelabs.conductor.Controller.detach(Controller.java:765)
    at com.bluelinelabs.conductor.Controller.destroy(Controller.java:881)
    at com.bluelinelabs.conductor.Controller.activityDestroyed(Controller.java:715)
    at com.bluelinelabs.conductor.Router.onActivityDestroyed(Router.java:424)
    at com.bluelinelabs.conductor.ActivityHostedRouter.onActivityDestroyed(ActivityHostedRouter.java:40)
    at com.bluelinelabs.conductor.internal.LifecycleHandler.destroyRouters(LifecycleHandler.java:161)
    at com.bluelinelabs.conductor.internal.LifecycleHandler.onDestroy(LifecycleHandler.java:130)
    at android.app.Fragment.performDestroy(Fragment.java:2433)
    at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1098)
    at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1148)
    at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1130)
    at android.app.FragmentManagerImpl.dispatchDestroy(FragmentManager.java:1981)
    at android.app.FragmentController.dispatchDestroy(FragmentController.java:218)
    at android.app.Activity.performDestroy(Activity.java:6406)
    at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1142)
    at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3818)
    ... 9 more
```
Looking at the [code](https://github.com/bluelinelabs/Conductor/blob/develop/conductor/src/main/java/com/bluelinelabs/conductor/ControllerHostedRouter.java#L50) I think the way you are looping is unsafe for concurrency: I would suggest to either use a thread-safe data structure or add some synchronization logic around there (and possibly other methods that use that collection) 😃

**Comment by EricKuck** [comment]
Surprised we hadn't come across this one yet. There was a bit of a race condition there, which is now fixed in the latest snapshot. Thanks!

**Comment by dmarcato** [comment]
Great, thanks! 👍 
Any timeline for a main release?

**Comment by EricKuck** [comment]
I'd like to fix at least #124 first. Hopefully in the next few days.

**Comment by EricKuck** [comment]
New version should be live now!
